### PR TITLE
make: add termdeps target

### PR DESCRIPTION
**Pull request by kaspar030**
### Contribution description

Previously, tests in CI on RasPis are run using "make flash-only term".
If "make term" requires work (e.g., compiling ethos), this makes "make term" startup long enough to possibly lose input, because "make reset" is called by "make test" with a fixed delay.

This PR introduces the termdeps target and makes use of it within the murdock scripts.
Test runs are now two-stage:

1. make flash-only termdeps -j2
2. make test

This a) lowers total test time if "make term" has work to do, by building the deps in parallel to flashing, and b) ensures that "make term" dependencies are already built when "make test" calls "make term".

### Testing procedure

AFAIK currently no test is using ethos, thus ideally, nothing will change for now.

### CODE REVIEW

**Comment by cladmi** [comment]
You could also simply add `TERMDEPS` to `TEST_DEPS`.
This way there are no other changes required.

**Comment by kaspar030** [comment]
> You could also simply add `TERMDEPS` to `TEST_DEPS`.
> This way there are no other changes required.

That would not decouple the "make flash-only termdeps", which can be done in parallel, from "make test". I'd like to keep "make test" being called without ".j".

**Comment by kaspar030** [comment]
> is there a way to get better evidence of the speed up?

Difficult. The tests you've mentioned are not run by the CI because they depend on scapy, which is not (yet) available there.

You can try doing what murdock would do (from ```$(RIOTBASE)```:
    $ ./.murdock run_test tests/gnrc_sock_dns samr21-xpro:gnu

But that fails for me for other reasons.

**Comment by cladmi** [comment]
> I'd like to keep "make test" being called without ".j".

Why ?

However, if for working, it is required that everything is compiled before doing `make test` then the `TERMDEPS` should also be added to`TEST_DEPS`.

Building it in parallel would just be a murdock optimization.

**Comment by kaspar030** [comment]
> However, if for working, it is required that everything is compiled before doing `make test` then the `TERMDEPS` should also be added to`TEST_DEPS`.
> 
> Building it in parallel would just be a murdock optimization.

"for working", everything is fine, in theory. The dependencies are sound. "make test" runs "make term", which in turn triggers the build of any possibly available "term" dependencies. Adding them to "TEST_DEPS" is cutting it short for one other slight flaw: "make term"'s startup time is only approximated by a variable (```TESTRUNNER_START_DELAY```, which defaults to three seconds). Unfortunately, using plain make dependencies, it is difficult to specify timing requirements within the dependencies.

So in order to not mess up the dependency graph for something we cannot express there, this PR exposes part of the graph as "termdeps", and makes murdock just make sure that has already been built before calling "make test".

> Why ?

No specific reason, other that I'd like a clearly define baseline before running "make test".

**Comment by kaspar030** [comment]
IRL talk with @cladmi, added the term deps to TEST_DEPS.

**Comment by fjmolinas** [comment]
> IRL talk with @cladmi, added the term deps to TEST_DEPS.

Ok in that case the second commit should be in another PR, since its a murdock speed up and since you are referencing the PR in the code I would rather have the referenced changed properly isolated. 

Otherwise it looks good and works for me locally. I guess @cladmi is agreeing with this since you talked IRL?

**Comment by kaspar030** [comment]
> Ok in that case the second commit should be in another PR, since its a murdock speed up and since you are referencing the PR in the code I would rather have the referenced changed properly isolated.

PR splitting is for making reviews easier, not for keeping changes distinct. Commits are used for the latter. AFAICT, the changes here are already agreed on. So unless you're having problems ACKing parts of this PR, splitting the PR is only extra work.

File: Makefile.include [hunk 1]
```diff
@@ -559,7 +559,10 @@ flash-only: $(FLASHDEPS)
 preflash: $(BUILD_BEFORE_FLASH)
 	$(PREFLASHER) $(PREFFLAGS)
 
-term: $(filter flash, $(MAKECMDGOALS)) $(TERMDEPS)
+termdeps: $(TERMDEPS)
+	@true
```
Comment (cladmi) [inline Makefile.include]: This is not required.
```suggestion
```

File: Makefile.include [hunk 2]
```diff
@@ -559,7 +559,10 @@ flash-only: $(FLASHDEPS)
...
+
+term: $(filter flash, $(MAKECMDGOALS)) termdeps
```
Comment (cladmi) [inline Makefile.include]: I would prefer to keep `TERMDEPS` here as there is not reason to depend on the named target.

File: Makefile.include [hunk 3]
```diff
@@ -588,6 +591,16 @@ reset:
...
                         $(shell test -f $(file) -a -x $(file) && echo $(file)))
+
+# "make test" calls "make term" which would implicitly build it's dependencies,
+# but that increases the time "make test" needs to get ready. That can cause
+# problems ("make term" missing some lines) as terminal startup is not properly
+# sychronized, but depends on a static timeout (TESTRUNNER_START_DELAY).
+# Murdock builds the term dependencies before running "make test" to circumvent
+# this. In order to make local builds behave similar, add the term deps here.
+# See #11762.
+TEST_DEPS += termdeps
```
Comment (cladmi) [inline Makefile.include]: I would prefer to keep `TERMDEPS` here too as there is not reason to depend on the named target.

**Comment by cladmi** [comment]
> > Ok in that case the second commit should be in another PR, since its a murdock speed up and since you are referencing the PR in the code I would rather have the referenced changed properly isolated.
> 
> PR splitting is for making reviews easier, not for keeping changes distinct. Commits are used for the latter. AFAICT, the changes here are already agreed on. So unless you're having problems ACKing parts of this PR, splitting the PR is only extra work.

The major change for me is to put `TERMDEPS` as `TEST_DEPS` and it is justified enough. As long as this one is in a separate commit it is good enough.
The `murdock` change will test itself.

Please squash.

**Comment by cladmi** [comment]
A cleanup for after this one, is to remove the `ethos` and others phony test-deps.
As right now, the compilation is re-tried every time.

**Comment by kaspar030** [comment]
- squashed

**Comment by kaspar030** [comment]
@cladmi this needs an ACK.

**Comment by cladmi** [comment]
I was AFK, and I also needed to do a final test for the `TERM_DEPS` so not possible to rush this. I will do it.

**Pull request merged**